### PR TITLE
Add D-Bus method to return Stratis pool-level metadata

### PR DESCRIPTION
**Pull request by mulkieran**
Related https://github.com/stratis-storage/project/issues/692

### CODE REVIEW

**Comment by mulkieran** [comment]
I'm not convinced that it would be wrong to have a parameter to modify what the method returns. For example, what if in future we do want to get what the metadata on the devices is and compare it with the metadata to write. We do not need to implement this right now, but we might want the flexibility to add it without adding another method.

**Review by jbaublitz — CHANGES_REQUESTED** [review]
I had a mix of requests and discussion points, but other than that this looks reasonable to me.

File: src/engine/strat_engine/backstore/blockdev.rs [hunk 1]
```diff
@@ -254,6 +254,27 @@ impl StratBlockDev {
...
+        let mut f = OpenOptions::new()
+            .read(true)
+            .open(self.underlying_device.metadata_path())?;
+        let header = static_header(&mut f)?.ok_or_else(|| {
+            StratisError::Msg("Stratis device has no signature buffer".to_string())
+        })?;
+        let bda = BDA::load(header, &mut f)?
+            .ok_or_else(|| StratisError::Msg("Stratis device has no BDA".to_string()))?;
+
+        self.bda = bda;
```
[the remark this answers is not in the document]
Comment (mulkieran) [reply src/engine/strat_engine/backstore/blockdev.rs:267]: This is a really important point. The last update time for the pool-level metadata for a particular device can not be determined unless you re-read the BDA. And if you re-read the BDA, then it seems like you ought to cache it. Another, probably better solution, is to cache the BDA when the pool-level metadata is written. If that seems like a better solution, I can try that separately. If that were done, then the cached BDA would be able to be used in this PR and the simplifications that you point out would be available.

[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/engine/strat_engine/backstore/blockdev.rs:267]: I think I'd prefer refreshing when the pool level metadata is written. I checked our existing usage of the bda, and I don't think we have a bug currently, but I was unaware of this side effect, and I think we're asking for trouble if we don't keep our cached data structures up to date as soon as they're written.

**Comment by mulkieran** [comment]
@jbaublitz https://github.com/stratis-storage/stratisd/pull/3606 which caches the BDA is up for your review.

File: src/dbus_api/pool/pool_3_7/api.rs [hunk 2]
```diff
@@ -19,3 +22,14 @@ pub fn destroy_filesystems_method(
...
         .out_arg(("return_string", "s"))
 }
+
+pub fn get_metadata_method(f: &Factory<MTSync<TData>, TData>) -> Method<MTSync<TData>, TData> {
+    f.method("Metadata", (), metadata)
+        .in_arg(("current", "b"))
+        // An array of bytes representing the pool-level metadata
+        //
+        // Rust representation: Vec<u8>
+        .out_arg(("results", "ay"))
```
[the remark this answers is not in the document]
Comment (mulkieran) [reply src/dbus_api/pool/pool_3_7/api.rs:32]: I feel like returning it as String w/out check is actually wrong. So, I'll try returning as String, but error in the unlikely case that it is not valid utf-8.

**Review by jbaublitz — APPROVED** [review]

**Pull request merged**